src/activity_item_handling.cpp

```diff
@@ -932,433 +939,1123 @@ static construction check_build_pre( const construction &con )
 {
-    zone_manager &mgr = zone_manager::get_manager();
-
-    const tripoint abspos = g->m.getabs( p.pos() );
-    const std::unordered_set<tripoint> &src_set = mgr.get_near(
-                zone_type_id( "CONSTRUCTION_BLUEPRINT" ), abspos );
-
-    const std::vector<tripoint> &src_sorted = get_sorted_tiles_by_distance( abspos, src_set );
-    const activity_id act_multiple_construction = activity_id( "ACT_BLUEPRINT_CONSTRUCTION" );
-
-    // Nuke the current activity, leaving the backlog alone
-    p.activity = player_activity();
-
-    // sort source tiles by distance
-    for( const tripoint &src : src_sorted ) {
-        const tripoint &src_loc = g->m.getlocal( src );
-        // check if somebodies already started it
-        partial_con *nc = g->m.partial_con_at( src_loc );
-        if( nc ) {
-            continue;
-        }
+    return std::all_of( con.required_skills.begin(), con.required_skills.end(),
+    [&]( const std::pair<skill_id, int> &pr ) {
+        return p->get_skill_level( pr.first ) >= pr.second;
+    } );
+}
 
-        if( !g->m.inbounds( src_loc ) ) {
-            if( !g->m.inbounds( p.pos() ) ) {
-                // p is implicitly an NPC that has been moved off the map, so reset the activity
-                // and unload them
-                p.assign_activity( act_multiple_construction );
-                p.set_moves( 0 );
-                g->reload_npcs();
-                return;
+static bool are_requirements_nearby( const std::vector<tripoint> &loot_spots,
+                                     const requirement_id &needed_things, const player &p, const activity_id activity_to_restore,
+                                     bool in_loot_zones )
+{
+    zone_manager &mgr = zone_manager::get_manager();
+    inventory temp_inv;
+    units::volume volume_allowed = p.volume_capacity() - p.volume_carried();
+    units::mass weight_allowed = p.weight_capacity() - p.weight_carried();
+    for( const tripoint &elem : loot_spots ) {
+        // if we are searching for things to fetch, we can skip certain thngs.
+        // if, however they are already near the work spot, then the crafting / inventory fucntions will have their own method to use or discount them.
+        if( in_loot_zones ) {
+            // skip tiles in IGNORE zone and tiles on fire
+            // (to prevent taking out wood off the lit brazier)
+            // and inaccessible furniture, like filled charcoal kiln
+            if( mgr.has( zone_type_id( "LOOT_IGNORE" ), g->m.getabs( elem ) ) ||
+                g->m.dangerous_field_at( elem ) ||
+                !g->m.can_put_items_ter_furn( elem ) ) {
+                continue;
             }
-            const std::vector<tripoint> route = route_adjacent( p, src_loc );
-            if( route.empty() ) {
-                // can't get there, can't do anything, skip it
+        }
+        for( const auto elem2 : g->m.i_at( elem ) ) {
```

Comment:
```suggestion
        for( const auto &elem2 : g->m.i_at( elem ) ) {
```
Copying the item is just going to slow the game down.